clients/drcachesim/tools/trace_filter.cpp

```diff
@@ -0,0 +1,217 @@
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include <fstream>
+#include <thread>
+#include "directory_iterator.h"
+#include "reader/trace_entry_file_reader.h"
+#include "trace_entry.h"
+#ifdef HAS_ZLIB
+#    include "reader/compressed_file_reader.h"
+#    include "common/gzip_ostream.h"
+#endif
+#include "trace_filter.h"
+
+typedef trace_entry_file_reader_t<std::ifstream> default_trace_entry_file_reader_t;
+
+trace_filter_t::trace_filter_t(const std::string &trace_dir,
+                               const std::string &output_dir, int worker_count,
+                               int verbosity)
+    : verbosity_(verbosity)
+    , worker_count_(worker_count)
+    , trace_dir_(trace_dir)
+    , output_dir_(output_dir)
+{
+    if (!init_file_reader_writer(trace_dir, output_dir, verbosity))
+        error_string_ = "Could not initialize trace reader and writer";
+    UNUSED(verbosity_);
+    UNUSED(output_prefix_);
+}
+
+#if defined(HAS_ZLIB)
+static bool
+ends_with(const std::string &str, const std::string &with)
+{
+    size_t pos = str.rfind(with);
+    if (pos == std::string::npos)
+        return false;
+    return (pos + with.size() == str.size());
+}
+#endif
+
+std::unique_ptr<trace_entry_reader_t>
+trace_filter_t::get_reader(const std::string &path, int verbosity)
+{
+#ifdef HAS_ZLIB
+    if (ends_with(path, trace_filter_t::gzip_suffix)) {
+        VPRINT(this, 3, "Using the gzip reader for %s\n", path.c_str());
+        input_file_format = FILE_FORMAT_GZIP;
+        return std::unique_ptr<trace_entry_reader_t>(
+            new compressed_trace_entry_file_reader_t(path, verbosity));
+    }
+#endif
+    VPRINT(this, 3, "Using the default reader for %s\n", path.c_str());
+    input_file_format = FILE_FORMAT_UNKNOWN;
+    return std::unique_ptr<trace_entry_reader_t>(
+        new default_trace_entry_file_reader_t(path, verbosity));
+}
+
+std::unique_ptr<std::ostream>
+trace_filter_t::get_writer(const std::string &path, int verbosity)
+{
+    std::unique_ptr<std::ostream> ofile;
+#ifdef HAS_ZLIB
+    if (input_file_format == FILE_FORMAT_GZIP) {
```

Comment:
This format var is not used anywhere else.  Maybe the reader and writer should be created at the same time to avoid the need to remember this in a separate routine.

Reply:
Removed.